### PR TITLE
CVE-2010-4035 and CVE-2016-5142

### PR DESCRIPTION
**Pull request by jmiller656**
Worked on these CVES:
CVE-2010-4035
CVE-2016-5142

### CODE REVIEW

File: cves/CVE-2010-4035.yml [hunk 1]
```diff
@@ -15,7 +15,10 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
   expressions" is fine, and security phrases like "invalid write" are fine to
   keep too.
-description: 
+description: |
+  There was a bug in the autofill feature which could allow attackers
```
Comment (paulhulbert) [inline cves/CVE-2010-4035.yml]: Might want to explicitly define what the bug is here.  Looking at the rest of the document it seems to be some sort of pointer/memory management issue, but it'd probably be good to outline it here.

File: cves/CVE-2010-4035.yml [hunk 2]
```diff
@@ -15,7 +15,10 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
```
Comment (paulhulbert) [inline cves/CVE-2010-4035.yml]: Any CWE?

**Comment by andymeneely** [comment]
Need the PR title to have the CVEs in them. Don't need your name - that's in the commit anyway.

File: cves/CVE-2010-4035.yml [hunk 3]
```diff
@@ -93,10 +106,10 @@ interesting_commits:
     interesting in light of the lessons learned from this vulnerability. Any
     emerging themes?
   commits:
-  - commit: 
-    note: 
-  - commit: 
-    note: 
+  - commit: dfa7be2725cd81ffd8f11a6b902159cde7db4957
+    note: An initial attempt at fixing the bug, it didn't work. "Contact me if you reach this condition".
```
Comment (paulhulbert) [inline cves/CVE-2010-4035.yml]: What did this commit do to attempt to fix the bug that didn't actually work?

File: cves/CVE-2016-5142.yml [hunk 4]
```diff
@@ -15,7 +15,10 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
```
Comment (paulhulbert) [inline cves/CVE-2016-5142.yml]: Any CWE?

File: cves/CVE-2016-5142.yml [hunk 5]
```diff
@@ -73,20 +79,23 @@ discovered:
 
     If there is no evidence as to how this vulnerability was found, then you may
     leave this part blank.
-  answer: 
-  date: 
-  automated: 
-  google: 
-  contest: 
+  answer: |
+    The vulnerability was discovered by an anonymous external researcher, who was
```
Comment (paulhulbert) [inline cves/CVE-2016-5142.yml]: Kinda a low hanging fruit, but there shouldn't be a comma after researcher.

File: cves/CVE-2016-5142.yml [hunk 6]
```diff
@@ -173,4 +184,11 @@ mistakes:
...
     engineering industry would find interesting.
-  answer: 
+  answer: |
+    The mistake that led to this vulnerability was the fact that google assumed
+    there were no vulnerabilities inside the web crypto library that they used in
+    chromium. Despite unit tests, there were vulnerabilities that went undiscovered
+    inside webcrypto. The lesson to be learned here is that you can't necessarily
+    trust every external piece of software that your system depends on. Luckily, 
+    regression of this bug is unlikely to occur because more unit tests were created
+    to test for the case that causes it.
```
Comment (paulhulbert) [inline cves/CVE-2016-5142.yml]: Do you know if they did anything to research the crypto library to make sure there weren't any other bugs lying around?

File: cves/CVE-2016-5142.yml [hunk 7]
```diff
@@ -15,7 +15,10 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
   expressions" is fine, and security phrases like "invalid write" are fine to
   keep too.
-description: 
+description: |
+  Due to poor handling of data buffers inside the cryptography API used in chrome
```
Comment (paulhulbert) [inline cves/CVE-2016-5142.yml]: Is this a buffer overflow?  If not, might want to clarify.

File: cves/CVE-2010-4035.yml [hunk 8]
```diff
@@ -31,8 +34,11 @@ fixes_vcc_instructions: |
...
 - :commit: 0d04639df7a3b2425c671ab08f68eb32b3d04cb1
-  :note: ''
-vccs: []
+  :note: |
+    The fix involved being more careful with handling of pointers, there
+    ended up being a pointer that was referenced after it was freed, causing
+    the crash.
+vccs: [cf8ebbba224c7d01170a46d81b3feda976298b67,74a5fec1709b6c1621ced52ddecf54004d8d09c0]
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
```
Comment (paulhulbert) [inline cves/CVE-2010-4035.yml]: 1 upvote

File: cves/CVE-2016-5142.yml [hunk 9]
```diff
@@ -34,7 +37,7 @@ fixes_vcc_instructions: |
 fixes:
 - :commit: 923aa7b0442d331e07b1853acd4f40d3a99a35ce
   :note: ''
-vccs: []
+vccs: [6c8a5ef266e1cb50ec3310030a818c51abc5af62]
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
```
Comment (paulhulbert) [inline cves/CVE-2016-5142.yml]: 1 upvote

File: cves/CVE-2010-4035.yml (hunk 2 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: I could not find a CWE for this issue

File: cves/CVE-2010-4035.yml (hunk 8 above)
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: 1 upvote

File: cves/CVE-2010-4035.yml [hunk 10]
```diff
@@ -71,20 +81,23 @@ discovered:
...
     What subsystems was the mistake in?
 
     Look at the path of the source code files code that were fixed to get
     directory names. Look at comments in the code. Look at the bug reports how
     the bug report was tagged.
-  answer: 
-  name: 
+  answer: |
+    The mistake was inside the autofill and translation subsystems.
+  name: autofill, translate
```
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: 1 up vote

File: cves/CVE-2010-4035.yml (hunk 3 above)
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: What do you mean by "Contact me if you reach this condition"?
Was this spoken by chromium team?

File: cves/CVE-2010-4035.yml [hunk 11]
```diff
@@ -171,4 +184,9 @@ mistakes:
     Use those questions to inspire your answer. Don't feel obligated to answer
     every one. Write a thoughtful entry here that those ing the software
     engineering industry would find interesting.
-  answer: 
+  answer: |
```
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: 1 up vote, but feels like the top of the line where you describe the vulnerability should be the description of the YML file tho

File: cves/CVE-2016-5142.yml (hunk 7 above)
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: This description definitely sounds like buffer overflow to me?

File: cves/CVE-2016-5142.yml (hunk 9 above)
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: 1 upvote

File: cves/CVE-2010-4035.yml [hunk 12]
```diff
@@ -53,9 +59,13 @@ unit_tested:
 
     For the "fix" answer below, check if the fix for the vulnerability involves
     adding or improving an automated test to ensure this doesn't happen again.
-  answer: 
-  code: 
-  fix: 
+  answer: |
+   The original code was not unit tested, and the fix does not involve unit tests.
```
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: 1 up vote

File: cves/CVE-2016-5142.yml [hunk 13]
```diff
@@ -55,9 +58,12 @@ unit_tested:
 
     For the "fix" answer below, check if the fix for the vulnerability involves
     adding or improving an automated test to ensure this doesn't happen again.
-  answer: 
-  code: 
-  fix: 
+  answer: |
+    The original code for this section was unit tested. Following discovery and
```
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: 1 up vote

File: cves/CVE-2016-5142.yml (hunk 5 above)
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: look closer, you might be able to find the username of the anonymous who got rewarded

File: cves/CVE-2016-5142.yml [hunk 14]
```diff
@@ -73,20 +79,23 @@ discovered:
...
     What subsystems was the mistake in?
 
     Look at the path of the source code files code that were fixed to get
     directory names. Look at comments in the code. Look at the bug reports how
     the bug report was tagged.
-  answer: 
-  name: 
+  answer: |
+    The mistake was inside the webkit crypto library.
+  name: Webkit
```
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: 1 up vote

File: cves/CVE-2016-5142.yml [hunk 15]
```diff
@@ -135,8 +144,10 @@ lessons:
     applies: 
     note: 
   frameworks_are_optional:
-    applies: 
-    note: 
+    applies: true
+    note: |
```
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: 1 up vote

File: cves/CVE-2010-4035.yml (hunk 2 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: Never mind, found one that made sense

File: cves/CVE-2010-4035.yml (hunk 11 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: I saw you wrote 1 upvote on this earlier. Is this a total of 2 upvotes for this file?

File: cves/CVE-2016-5142.yml (hunk 4 above)
Comment (jmiller656) [inline cves/CVE-2016-5142.yml]: Found one

File: cves/CVE-2010-4035.yml (hunk 3 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: There was a comment in the code that said "contact me"
I ended up fixing this part to make it more clear

File: cves/CVE-2016-5142.yml (hunk 5 above)
Comment (jmiller656) [inline cves/CVE-2016-5142.yml]: Whoops. Still wasn't able to find username of the anon. The related bug is still blocked from public view

File: cves/CVE-2016-5142.yml (hunk 6 above)
Comment (jmiller656) [inline cves/CVE-2016-5142.yml]: Unfortunately, I didn't see anything involving that

File: cves/CVE-2016-5142.yml (hunk 7 above)
Comment (jmiller656) [inline cves/CVE-2016-5142.yml]: This actually wasn't a buffer overflow, which indicated my description of the bug was not very good. I've made it more specific now

Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: sounds better now

File: cves/CVE-2010-4035.yml (hunk 11 above)
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: sorry for confusion. I meant to up vote for each line and hoping for you to count all up, but @paulhulbert mentioned I shouldn't do that. I will fix up vote count at the end of the each yml file

File: cves/CVE-2010-4035.yml [hunk 16]
```diff
@@ -171,4 +199,9 @@ mistakes:
     Use those questions to inspire your answer. Don't feel obligated to answer
     every one. Write a thoughtful entry here that those ing the software
     engineering industry would find interesting.
-  answer: 
+  answer: |
+    This problem seems to be caused by a coding mistake. For both the subsystems
+    that the vulnerability existed in, the problem involved incorrect usage of 
+    pointers. Low level languages where memory management has to be done manually
+    can be tricky, it would be shocking if no such errors ever occurred inside a
+    project as large as chrome.
```
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: 4 up vote

File: cves/CVE-2016-5142.yml [hunk 17]
```diff
@@ -173,4 +190,11 @@ mistakes:
     Use those questions to inspire your answer. Don't feel obligated to answer
     every one. Write a thoughtful entry here that those ing the software
     engineering industry would find interesting.
-  answer: 
+  answer: |
```
Comment (hinza0188) [inline cves/CVE-2016-5142.yml]: 5 up vote

File: cves/CVE-2010-4035.yml (hunk 16 above)
Comment (hinza0188) [inline cves/CVE-2010-4035.yml]: 4 up vote

File: cves/CVE-2016-5142.yml (hunk 7 above)
Comment (jmiller656) [inline cves/CVE-2016-5142.yml]: Awesome! Thanks for the feedback.

File: cves/CVE-2010-4035.yml [hunk 18]
```diff
@@ -15,7 +15,11 @@ description_instructions: |
   that outsiders to Chromium would not understand. Technology like "regular
   expressions" is fine, and security phrases like "invalid write" are fine to
   keep too.
-description: 
+description: |
+  There was a bug in the autofill and translation features which could allow attackers
+  to cause the application to crash remotely, therefore introducing a
```
Comment (andymeneely) [inline cves/CVE-2010-4035.yml]: I could use a little bit more detail on this one. This is a good opening paragraph. Second paragraph explain a little more about the bug to someone who has some programming experienc.

File: cves/CVE-2010-4035.yml [hunk 19]
```diff
@@ -31,16 +35,27 @@ fixes_vcc_instructions: |
   CVE-2011-3092.yml). Fixes and VCCs follow the same format.
 fixes:
 - :commit: 0d04639df7a3b2425c671ab08f68eb32b3d04cb1
-  :note: ''
-vccs: []
+  :note: |
+    The fix involved being more careful with handling of pointers, there
+    ended up being a pointer that was referenced after it was freed, causing
+    the crash.
```
Comment (andymeneely) [inline cves/CVE-2010-4035.yml]: I would also note that the fix involved changing variable types too. And deleting code in some places, which is an interesting choice. There might be some low-level design choices they made in addition to coding.

File: cves/CVE-2010-4035.yml [hunk 20]
```diff
@@ -31,16 +35,27 @@ fixes_vcc_instructions: |
...
+vccs:
+- :commit: cf8ebbba224c7d01170a46d81b3feda976298b67
+  :note: |
+    This commit introduced the section of the code where the inforbar delegate
+    pointer was used irresponsibly inside the credit card autofil system
```
Comment (andymeneely) [inline cves/CVE-2010-4035.yml]: "infobar"

File: cves/CVE-2010-4035.yml [hunk 21]
```diff
@@ -31,16 +35,27 @@ fixes_vcc_instructions: |
...
+    pointer was used irresponsibly inside the translation system
 upvotes_instructions: |
   For the first round, ignore this upvotes number.
 
   For the second round of reviewing, you will be giving a certain amount of
   upvotes to each vulnerability you see. Your peers will tell you how
   interesting they think this vulnerability is, and you'll add that to the
   upvotes score on your branch.
-upvotes: 
+upvotes: 5
```
Comment (andymeneely) [inline cves/CVE-2010-4035.yml]: +2 from me

File: cves/CVE-2010-4035.yml (hunk 20 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: Whoops, got it.

File: cves/CVE-2010-4035.yml (hunk 19 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: Oh, that's true. Added an explanation of what they changed and why.

File: cves/CVE-2010-4035.yml (hunk 18 above)
Comment (jmiller656) [inline cves/CVE-2010-4035.yml]: Good point, I added a more in-depth explanation for the second paragraph.

**Pull request merged**